stereo_points/stereo_points.py

```diff
@@ -107,205 +104,194 @@ def setup(self):
+        if (newNode is not None) and (self.GetCoordTable() is not None):  # to not run the updatepoints when nothing is selected
+            self.updatePointsCoordsFromXYZ(self.GetCoordTable(), self.referenceImage_selectionCombo.currentNode(), newNode)
 
     def onControlPointSelectedChanged(self, newNode):
-        # print('selection changed !')
-        if (type(newNode) == type(slicer.vtkMRMLMarkupsLineNode())):
+        # for each markupLineNode, we maintain a table with the coord conversion.
+        # table nodes are names with the name of the markupLine with _coordsConversion as prefix
+        if type(newNode) == type(slicer.vtkMRMLMarkupsLineNode()):
             coordTableName = newNode.GetName() + "_coordsConversion"
         else:
             return
-        # print('looking for '+coordTableName)
-
-        if slicer.mrmlScene.GetNodesByName(coordTableName).GetNumberOfItems() == 0:
-            # print('create new table')
-            coordTable = slicer.vtkMRMLTableNode()
-            coordTable.SetName(coordTableName)
-            coordTable.SetLocked(True)
-            for col in ['Marker', 'x', 'y', 'z', 'r', 'a', 'd', 'X', 'Y', 'Z', 'R', 'A', 'S', 'i', 'j', 'k']:
-                c = coordTable.AddColumn()
+
+        coordTable_id = newNode.GetNodeReferenceID("stereotaxia_coordTable")
+
+        # if the lineMarkup does not reference any coordtable yet, create it
+        if coordTable_id is None:
+            # prepare a new table
+            newCoordTable = slicer.vtkMRMLTableNode()
+            newCoordTable.SetName(coordTableName)
+            newCoordTable.SetLocked(True)
+            # columns:
+            # x, y, z, r, a, d: leksell frame settings + depth
+            # XYZ cartesian coordinates in leksell space
+            # RAS coordinates in physical space (what slicer uses to place the points)
+            # ijk coordinates in image space
+            for col in ["Marker", "x", "y", "z", "r", "a", "d", "X", "Y", "Z", "R", "A", "S", "i", "j", "k"]:
+                c = newCoordTable.AddColumn()
                 c.SetName(col)
 
-            slicer.mrmlScene.AddNode(coordTable)
-            # keep a map of the fiducialNode <-> TableNode couples
-            self.fiducialNodesToCoordTables_id_tuples.append((newNode.GetID(), coordTable.GetID()))
+            slicer.mrmlScene.AddNode(newCoordTable)
+            # refererence the line from the table so we can get back to it
+            newCoordTable.AddNodeReferenceID("stereotaxia_trajLine", newNode.GetID())
+            newNode.AddNodeReferenceID("stereotaxia_coordTable", newCoordTable.GetID())
             # add an observer for both nodes, whenever one is renamed, the other one is renamed as well
-            coordTable.AddObserver(vtk.vtkCommand.ModifiedEvent, self.onCoordTableModified)
+            self.lObs_tableNode = newCoordTable.AddObserver(vtk.vtkCommand.ModifiedEvent, self.onCoordTableModified)
```

Comment:
For each `AddObserver` there must be a matching `RemoveObserver`.
It is great that you save `self.lObs_tableNode`, now you just need to make sure to remove this observer:
- before adding a new observer to another table node (e.g., when you select a different fiducial group)
- in the `cleanup` method (to avoid memory leaks when you exit the application)

To make it easier to remove an observer, I would recommend to save the observed node name as well. For example:

```python
self.tableNodeObservation = [newCoordTable, newCoordTable.AddObserver(vtk.vtkCommand.ModifiedEvent, self.onCoordTableModified)]
```

How to remove the observer:

```python
if self.tableNodeObservation:
    self.tableNodeObservation[0].RemoveObserver(self.tableNodeObservation[1])
    self.tableNodeObservation = None
```

Reply:
Thanks, this is helpful. We don't want to remove the observer when the user selects another markupline in the combobox. But we still need to save all the observers and cleanup. So I quess tableNodeObservation will be a list to store all observers for cleanup.